File: src/main/problems/java/Nr39.java
package main.problems.java;

import java.util.ArrayList;
import java.util.LinkedList;
import java.util.List;

public class Nr39 {
    public static void main(String[] args) {
        Nr39 nr39 = new Nr39();
        nr39.combinationSum(new int[]{2, 3, 6, 7}, 7);
    }

    public List<List<Integer>> combinationSum(int[] candidates, int target) {
        return combinationSum(
                candidates,
                target,
                0,
                new LinkedList<>(),
                0,
                new ArrayList<>()
        );
    }

    private List<List<Integer>> combinationSum(
            int[] candidates,
            int target,
            int index,
            LinkedList<Integer> currentList,
            int currentSum,
            List<List<Integer>> result
    ) {
        if (currentSum > target || index >= candidates.length) {
            return result;
        }
        if (currentSum == target) {
            result.add(new ArrayList<>(currentList));
        }
        for (int i = index; i < candidates.length; i++) {
            currentList.add(candidates[i]);
            combinationSum(
                    candidates,
                    target,
                    i,
                    currentList,
                    currentSum + candidates[i],
                    result
            );
            currentList.removeLast();
        }
        return result;
    }
}

File: src/main/problems/ruby/Nr39.rb
# @param {Integer[]} candidates
# @param {Integer} target
# @return {Integer[][]}
def combination_sum(candidates, target, current_sum = 0, current = [], start = 0)
    return [current.dup] if current_sum == target
    return nil if current_sum > target
    return nil if start >= candidates.size

    # Take current
    current << candidates[start]
    res_take = combination_sum(
        candidates, 
        target, 
        current_sum + candidates[start], 
        current,
        start
    )
    current.pop

    # Skip current
    res_skip = combination_sum(
        candidates,
        target,
        current_sum,
        current,
        start + 1
    )

    (res_take || []) + (res_skip || [])
end